filebeat/input/filestream/copytruncate_prospector.go

```diff
@@ -0,0 +1,344 @@
+	return f.path
+}
+
+// rotatedFilestream includes the information of the original file
+// and its identifier, and the rotated file.
+type rotatedFilestream struct {
+	originalSrc loginp.Source
+	rotated     []rotatedFileInfo
+}
+
+func newRotatedFilestreams(cfg *copyTruncateConfig) *rotatedFilestreams {
+	var sorter sorter
+	sorter = &defaultSorter{}
+	if cfg.DateFormat != "" {
+		sorter = &dateSorter{cfg.DateFormat}
+	}
+	return &rotatedFilestreams{
+		table:  make(map[string]*rotatedFilestream, 0),
+		count:  cfg.Count,
+		sorter: sorter,
+	}
+}
+
+// defaultSorter sorts rotated log files that have a numeric suffix.
+// Example: apache.log.1, apache.log.2
+type defaultSorter struct{}
+
+func (s *defaultSorter) sort(files []rotatedFileInfo) {
+	sort.Slice(
+		files,
+		func(i, j int) bool {
+			return files[i].path < files[j].path
+		},
+	)
+}
+
+// dateSorter sorts rotated log files that have a date suffix
+// based on the configured format.
+// Example: apache.log-21210526, apache.log-20210527
+type dateSorter struct {
+	format string
+}
+
+func (s *dateSorter) sort(files []rotatedFileInfo) {
+	sort.Slice(
+		files,
+		func(i, j int) bool {
+			return s.GetTs(&files[i]).After(s.GetTs(&files[j]))
+		},
+	)
+}
+
+func (s *dateSorter) GetTs(fi *rotatedFileInfo) time.Time {
+	if !fi.ts.IsZero() {
+		return fi.ts
+	}
+	filename := filepath.Base(fi.path)
+	fileTs := (filename[len(filename)-len(s.format):])
+
+	ts, err := time.Parse(s.format, fileTs)
+	if err != nil {
+		return time.Time{}
+	}
+	fi.ts = ts
+	return ts
+}
+
+// rotatedFilestreams is a map of original files and their rotated instances.
+type rotatedFilestreams struct {
+	table  map[string]*rotatedFilestream
+	count  int
+	sorter sorter
+}
+
+// addOriginalFile adds a new original file and its identifying information
+// to the bookkeeper.
+func (r rotatedFilestreams) addOriginalFile(path string, src loginp.Source) {
+	if _, ok := r.table[path]; ok {
+		return
+	}
+	r.table[path] = &rotatedFilestream{originalSrc: src, rotated: make([]rotatedFileInfo, 0, r.count)}
+}
+
+// isOriginalAdded checks if an original file has been found.
+func (r rotatedFilestreams) isOriginalAdded(path string) bool {
+	_, ok := r.table[path]
+	return ok
+}
+
+// originalSrc returns the original Source information of a given
+// original file path.
+func (r rotatedFilestreams) originalSrc(path string) loginp.Source {
+	return r.table[path].originalSrc
+}
+
+// addRotatedFile adds a new rotated file to the list and returns its index.
+// if a file is already added, the source is updated and the index is returned.
+func (r rotatedFilestreams) addRotatedFile(original, rotated string, src loginp.Source) int {
+	for idx, fi := range r.table[original].rotated {
+		if fi.path == rotated {
+			r.table[original].rotated[idx].src = src
+			return idx
+		}
+	}
+
+	r.table[original].rotated = append(r.table[original].rotated, rotatedFileInfo{rotated, src, time.Time{}})
+	r.sorter.sort(r.table[original].rotated)
```

Comment:
TODO (for follow up PR?): Full sort per entry to be added might become quite expensive for users with a large number of files. Having a sort on insert might not be the fastest either, depending on the sort algorithm (sort package uses quicksort for 'big' arrays and insertion sort for small arrays I think). Instead of having a sorter consider an `appender` interface that might use `sort`, or allow for a more sophisticated implementation.